obspy/core/inventory/response.py

```diff
@@ -767,6 +768,124 @@ def __init__(self, resource_id=None, instrument_sensitivity=None,
+                  if _i.stage_sequence_number]
+        if not stages:
+            return {}
+
+        if list(range(1, len(stages) + 1)) != stages:
+            raise ValueError("Can only determine sampling rates if response "
+                             "stages are in order.")
+
+        # First fill in all the set values.
+        sampling_rates = {}
+        for stage in self.response_stages:
+            input_sr = None
+            output_sr = None
+            factor = None
+            if stage.decimation_input_sample_rate:
+                input_sr = stage.decimation_input_sample_rate
+                if stage.decimation_factor:
+                    factor = stage.decimation_factor
+                    output_sr = input_sr / float(factor)
+            sampling_rates[stage.stage_sequence_number] = {
+                "input_sampling_rate": input_sr,
+                "output_sampling_rate": output_sr,
+                "decimation_factor": factor}
+
+        # Nothing might be set - just return in that case.
+        if set(itertools.chain.from_iterable(v.values()
+               for v in sampling_rates.values())) == {None}:
+            return sampling_rates
+
+        # Find the first set input sampling rate. The output sampling rate
+        # cannot be set without it. Set all prior input and output sampling
+        # rates to it.
+        for i in stages:
+            sr = sampling_rates[i]["input_sampling_rate"]
+            if sr:
+                for j in range(1, i):
+                    sampling_rates[j]["input_sampling_rate"] = sr
+                    sampling_rates[j]["output_sampling_rate"] = sr
+                    sampling_rates[j]["decimation_factor"] = 1
+                break
+
+        # This should guarantee that the input and output sampling rate of the
+        # the first stage are set.
+        output_sr = sampling_rates[1]["output_sampling_rate"]
+        if not output_sr:  # pragma: no cover
+            raise NotImplementedError
+
+        for i in stages:
+            si = sampling_rates[i]
+            if not si["input_sampling_rate"]:
+                si["input_sampling_rate"] = output_sr
+            if not si["output_sampling_rate"]:
+                if not si["decimation_factor"]:
+                    si["output_sampling_rate"] = si["input_sampling_rate"]
+                    si["decimation_factor"] = 1
+                else:
+                    si["output_sampling_rate"] = si["input_sampling_rate"] / \
+                        float(si["decimation_factor"])
+            if not si["decimation_factor"]:
+                si["decimation_factor"] = int(round(
+                    si["input_sampling_rate"] / si["output_sampling_rate"]))
+            output_sr = si["output_sampling_rate"]
+
+        def is_close(a, b):
+            return abs(a - b) < 1e-5
+
+        # Final consistency checks.
+        sr = sampling_rates[stages[0]]["input_sampling_rate"]
```

Comment:
First check is against itself right? `sampling_rates[stages[0]]` seems identical to `sampling_rates[i]` on its first iteration. Doesn't really matter anyway

Reply:
Yes the first check is against itself - but later loop iterations are against the previous stage's output sampling rate. Is there a cleaner way to code this?

Reply:
Nah it's not gonna make it easier but I was just trying to follow the code paths